// @netlify/zip-it-and-ship-it/dist/runtimes/node/bundlers/nft/transpile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transpile = void 0;
const esbuild_1 = require("@netlify/esbuild");
const error_js_1 = require("../../../../utils/error.js");
const bundler_target_js_1 = require("../esbuild/bundler_target.js");
const transpile = async (path, config, functionName) => {
    // The version of ECMAScript to use as the build target. This will determine
    // whether certain features are transpiled down or left untransformed.
    const nodeTarget = (0, bundler_target_js_1.getBundlerTarget)(config.nodeVersion);
    try {
        const transpiled = await (0, esbuild_1.build)({
            bundle: false,
            entryPoints: [path],
            format: 'cjs',
            logLevel: 'error',
            platform: 'node',
            sourcemap: Boolean(config.nodeSourcemap),
            target: [nodeTarget],
            write: false,
        });
        return transpiled.outputFiles[0].text;
    }
    catch (error) {
        throw error_js_1.FunctionBundlingUserError.addCustomErrorInfo(error, { functionName, runtime: 'js', bundler: 'nft' });
    }
};
exports.transpile = transpile;
