// @netlify/zip-it-and-ship-it/dist/runtimes/go/builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.build = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const error_js_1 = require("../../utils/error.js");
const shell_js_1 = require("../../utils/shell.js");
const build = async ({ destPath, mainFile, srcDir }) => {
    const functionName = (0, path_1.basename)(srcDir);
    try {
        await shell_js_1.shellUtils.runCommand('go', ['build', '-o', destPath, '-ldflags', '-s -w'], {
            cwd: srcDir,
            env: {
                CGO_ENABLED: '0',
                GOOS: 'linux',
                GOARCH: 'amd64',
            },
        });
    }
    catch (error) {
        console.error(`Could not compile Go function ${functionName}:\n`);
        throw error_js_1.FunctionBundlingUserError.addCustomErrorInfo(error, { functionName, runtime: 'go' });
    }
    const stat = await fs_1.promises.lstat(destPath);
    return {
        mainFile,
        name: functionName,
        srcDir,
        srcPath: destPath,
        stat,
    };
};
exports.build = build;
